packages/optimizely-sdk/lib/core/audience_evaluator/index.js

```diff
@@ -16,54 +16,88 @@
+ * @return {Boolean}                                            true if the user attributes match the given audience conditions, false
+ *                                                              otherwise
+ */
+AudienceEvaluator.prototype.evaluate = function(audienceConditions, audiencesById, userAttributes) {
+  // if there are no audiences, return true because that means ALL users are included in the experiment
+  if (!audienceConditions || audienceConditions.length === 0) {
+    return true;
+  }
 
-    var evaluateAudience = function(audienceId) {
-      var audience = audiencesById[audienceId];
-      if (audience) {
-        logger.log(LOG_LEVEL.DEBUG, sprintf(LOG_MESSAGES.EVALUATING_AUDIENCE, MODULE_NAME, audienceId, JSON.stringify(audience.conditions)));
-        var result = conditionTreeEvaluator.evaluate(audience.conditions, evaluateConditionWithUserAttributes);
-        var resultText = result === null ? 'UNKNOWN' : result.toString().toUpperCase();
-        logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.AUDIENCE_EVALUATION_RESULT, MODULE_NAME, audienceId, resultText));
-        return result;
-      }
+  if (!userAttributes) {
+    userAttributes = {};
+  }
 
-      return null;
-    };
+  var evaluateAudience = function(audienceId) {
+    var audience = audiencesById[audienceId];
+    if (audience) {
+      logger.log(LOG_LEVEL.DEBUG, sprintf(LOG_MESSAGES.EVALUATING_AUDIENCE, MODULE_NAME, audienceId, JSON.stringify(audience.conditions)));
+      var result = conditionTreeEvaluator.evaluate(audience.conditions, this.evaluateConditionWithUserAttributes.bind(this, userAttributes));
+      var resultText = result === null ? 'UNKNOWN' : result.toString().toUpperCase();
+      logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.AUDIENCE_EVALUATION_RESULT, MODULE_NAME, audienceId, resultText));
+      return result;
+    }
 
-    return conditionTreeEvaluator.evaluate(audienceConditions, evaluateAudience) || false;
-  },
+    return null;
+  }.bind(this);
+
+  return conditionTreeEvaluator.evaluate(audienceConditions, evaluateAudience) || false;
+};
+
+/**
+ * Wrapper around evaluator.evaluate that is passed to the conditionTreeEvaluator.
+ * Evaluates the condition provided given the user attributes if an evaluator has been defined for the condition type.
+ * @param  {Object} userAttributes     A map of user attributes.
+ * @param  {Object} condition          A single condition object to evaluate.
+ * @return {Boolean|null}              true if the condition is satisfied, null if a matcher is not found.
+ */
+AudienceEvaluator.prototype.evaluateConditionWithUserAttributes = function(userAttributes, condition) {
+  var evaluator = this.typeToEvaluatorMap[condition.type];
+  if (!evaluator) {
+    logger.log(LOG_LEVEL.WARNING, sprintf(LOG_MESSAGES.UNKNOWN_CONDITION_TYPE, MODULE_NAME, JSON.stringify(condition)));
+    return null;
+  }
+  try {
+    return evaluator.evaluate(condition, userAttributes);
```

Comment:
I still think we need to think about the arguments we're passing here.

* Primarily we need to pass a logger or update the `custom_attribute_condition_evaluator` to import the logger directly
* Can we pass an object with named properties?

Reply:
This is a good point.

We **_DO_** need to pass a logger - I mistakenly omitted this. We can't have `custom_attribute_condition_evaluator` import it as its stateful to this instance of `AudienceEvaluator`

> Can we pass an object with named properties?
WDYM exactly? The `custom_attribute_condition_evaluator` needs the `userAttributes` map to do it's matching, but other matchers won't necessarily need them.

Reply:
Oh. I'm proposing named parameters like this: 
```suggestion
    return evaluator.evaluate({
      condition: condition,
      userAttributes: userAttributes,
    });
```
This is how we invoke customer's notification listener callbacks ([example](https://docs.developers.optimizely.com/full-stack/docs/register-notification-listeners#section-track-notification-listener)) and, arguably, is what we ought to support in all of our public API methods.

Reply:
Oh. I'm proposing named parameters like this: 
```suggestion
    return evaluator.evaluate({
      condition: condition,
      userAttributes: userAttributes,
    });
```
This is how we invoke customer's notification listener callbacks ([example](https://docs.developers.optimizely.com/full-stack/docs/register-notification-listeners#section-track-notification-listener)) and, arguably, is how we ought to accept inputs to all of our public API methods.

Reply:
Ok, doing that here would involve changing a lot of established code (like [this](https://github.com/optimizely/javascript-sdk/blob/master/packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.js)). I don't think it makes a ton of sense to make that change in this PR.

Reply:
Hmm, the changes ought to be limited to [`custom_attribute_condition_evaluator.evaluate`](https://github.com/optimizely/javascript-sdk/blob/bd49de091af215c867bb1017ec3b77337cc33b33/packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.js#L48-L77) and its unit tests. I think this is the best time to take this on - we're starting to move towards a public API, and we're prepare to introduce many more evaluators than `custom_attribute_condition_evaluator`